File: Javascript/2.Settembre/33.ScreenTime-12Settembre.js
//todo Dato un array di input di sette numeri interi, che rappresentano una settimana, dove ogni numero intero rappresenta la quantità di ore trascorse al telefono quel giorno, determina se il tempo trascorso davanti allo schermo è eccessivo in base a questi vincoli:

//! Se un singolo giorno ha 10 ore o più, è eccessivo.
//! Se la media di tre giorni consecutivi è maggiore o uguale a 8 ore, è eccessivo.
//! Se la media dei sette giorni è maggiore o uguale a 6 ore, è eccessivo.

//? Tempo di esecuzione:
//* 11 Minuti e 42 Secondi

//?Codice scritto senza utilizzo di ChatGpt

function tooMuchScreenTime(hours) {

  let totWeek = hours.reduce((acc, value) => acc + value, 0)
  let mediaWeek = totWeek / 7
  let isFalse = null

  for(let i = 0; i < hours.length; i++) {
    if(hours[i] >= 10 ||
      (hours[i] + hours[i+1] + hours[i+2]) / 3 >= 8 ||
      mediaWeek >= 6) return true;
  else isFalse = false
  }

  return isFalse;
}

console.log(tooMuchScreenTime([3, 3, 5, 8, 8, 9, 4]))

//?Codice scritto da ChatGpt
//* Nessun codice scritto da ChatGpt

//* ✅ PRO

//* Logica di base corretta: controlla le 3 condizioni

//* Usa reduce per sommare → bene

//* Funzione restituisce boolean

//! ❌ CONTRO

//! isFalse inutile → puoi return false direttamente

//! Il ciclo for con hours[i+1] e hours[i+2] può uscire dai limiti dell’array

//! Calcolo della media settimanale ripetuto ogni iterazione → inutile

//! Ciclo e condizioni più complicati del necessario

//? Versione Chat Gpt 

function tooMuchScreenTime(hours) {
  const mediaWeek = hours.reduce((a, b) => a + b, 0) / 7;

  if (mediaWeek >= 6) return true;

  for (let i = 0; i < hours.length; i++) {
    if (hours[i] >= 10) return true;
    if (i + 2 < hours.length) {
      const media3 = (hours[i] + hours[i + 1] + hours[i + 2]) / 3;
      if (media3 >= 8) return true;
    }
  }

  return false;
}

console.log(tooMuchScreenTime([3, 3, 5, 8, 8, 9, 4])); // false

//todo Considerazioni: 

//! N.B.

File: Python/2.Settembre/33.Screen_time-12Settembre.py
# //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio, codice e vorrei che tu mi dessi consigli

# //todo Consegna dell'esercizio

# Dato un array di sette numeri interi che rappresentano una settimana,
# dove ogni numero rappresenta le ore al telefono per giorno,
# determinare se il tempo è eccessivo in base a:
# - un singolo giorno > 10 ore
# - media di 3 giorni consecutivi >= 8 ore
# - media dei 7 giorni >= 6 ore

# //todo


# //*Codice scritto senza utilizzo di ChatGpt

from functools import reduce

def too_much_screen_time(hours):
    if reduce(lambda x, y: x + y, hours) / len(hours) >= 6:
        return True

    for i in range(len(hours)):
        time = hours[i];
        if time >= 10:
            return True;

        if i < len(hours) - 2:
            media_tre_giorni = (hours[i] + hours[i+1] + hours[i+2]) / 3
            if media_tre_giorni >= 8:
                return True
        
    return False

print(too_much_screen_time([1, 2, 3, 4, 5, 6, 7]))

# //*


# //* Codice scritto da ChatGpt

def too_much_screen_time(hours):
    if max(hours) > 10:
        return True

    if sum(hours) / 7 >= 6:
        return True

    for i in range(len(hours) - 2):
        if (hours[i] + hours[i+1] + hours[i+2]) / 3 >= 8:
            return True

    return False

# //*


# //* Pro e contro del mio codice

# //* Pro

# - La logica generale è corretta e rispetta tutti i vincoli.
# - Hai gestito bene i casi di media su 3 giorni consecutivi.
# - L'uso di reduce mostra buona conoscenza degli strumenti Python.
# - Il codice è chiaro nella separazione dei controlli.

# //! Contro

# - Il controllo della media dei 7 giorni viene fatto prima di controllare il singolo giorno (>10), mentre la consegna suggerisce priorità logica indipendente.
# - Usi reduce quando sum() sarebbe più semplice e leggibile.
# - Il ciclo controlla anche la media dopo aver già verificato il totale, rendendo alcune operazioni inutili.
# - I punti e virgola non sono necessari in Python.
# - Il loop parte da 0 fino a len(hours), ma per le triple conviene fermarsi prima.

# //*


# //* Consigli

# - Usa sum() invece di reduce: è più Pythonico e immediato.
# - Controlla prima i casi più semplici e veloci (max e sum), poi le condizioni più complesse.
# - Nel ciclo delle finestre di 3 giorni usa direttamente range(len(hours) - 2).
# - Quando lavori con finestre scorrevoli, pensa sempre a evitare controlli fuori indice.
# - Ottimo esercizio: stai già ragionando bene sulle condizioni multiple.

# Valutazione: 8.5/10
# Buona logica, ma si può migliorare in efficienza e pulizia del codice.